hail/python/hailtop/hailctl/config/utils.py

```diff
@@ -0,0 +1,149 @@
+import json
+import tempfile
+from typing import Dict, List, Optional
+
+
+class InsufficientPermissions(Exception):
+    def __init__(self, message: str):
+        self.message = message
+
+
+async def already_logged_into_service() -> bool:
+    from hailtop.auth.auth import async_get_userinfo  # pylint: disable=import-outside-toplevel
+    try:
+        await async_get_userinfo()
+        return True
+    except Exception:
+        return False
+
+
+async def check_for_gcloud() -> bool:
+    from hailtop.utils import check_shell  # pylint: disable=import-outside-toplevel
+    try:
+        await check_shell("gcloud version")
+        return True
+    except Exception:
+        return False
+
+
+async def get_gcp_default_project(verbose: bool) -> Optional[str]:
+    from hailtop.utils import check_shell_output  # pylint: disable=import-outside-toplevel
+    try:
+        project, _ = await check_shell_output("gcloud config get-value project", echo=verbose)
+        project = project.strip().decode('utf-8')
+        return project
+    except Exception:
+        return None
+
+
+def get_default_region(supported_regions: List[str], cloud: str) -> Optional[str]:
+    if cloud == 'gcp':
+        if 'us-central1' in supported_regions:
+            return 'us-central1'
+    elif cloud == 'azure':
+        if 'eastus' in supported_regions:
+            return 'eastus'
+    return None
+
+
+async def get_gcp_bucket_information(bucket: str, verbose: bool) -> Optional[dict]:
+    from hailtop.utils import CalledProcessError, check_shell_output  # pylint: disable=import-outside-toplevel
+    try:
+        info, _ = await check_shell_output(f'gcloud storage buckets describe gs://{bucket} --format="json"', echo=verbose)
+        return json.loads(info.decode('utf-8'))
+    except CalledProcessError as e:
+        if 'does not have storage.buckets.get access to the Google Cloud Storage bucket' in e.stderr.decode('utf-8'):
+            msg = f'ERROR: You do not have sufficient permissions to get information about bucket {bucket} or it does not exist. ' \
+                  f'If the bucket exists, ask a project administrator to assign you the StorageAdmin role in Google Cloud Storage.'
+            raise InsufficientPermissions(msg) from e
+        raise
+
+
+async def create_gcp_bucket(*,
+                            project: str,
+                            bucket: str,
+                            location: str,
+                            lifecycle_days: Optional[int],
+                            labels: Optional[Dict[str, str]],
+                            verbose: bool):
+    from hailtop.utils import CalledProcessError, check_shell  # pylint: disable=import-outside-toplevel
+    if labels:
+        labels_str = ','.join(f'{k}={v}' for k, v in labels.items())
+    else:
+        labels_str = None
+
+    try:
+        await check_shell(f'gcloud --project {project} storage buckets create gs://{bucket} --location={location}', echo=verbose)
```

Comment:
To use `check_shell` we need to shell quote, but this isn't really using any shell functionality just the `gcloud` executable, so this should use `check_exec_output`.

Reply:
I find writing commands for `check_exec_output` to be significantly harder to read.

Reply:
Fair point, but I think using the shell when we don't have to is a footgun because we could forget to properly quote inputs and shell invocations end up in bash history files (so we have to be careful not to `check_shell` with anything sensitive). What if we just made a wrapper for `check_exec_output` that takes a single string, splits on spaces, and passes it to `check_exec_output`?

Reply:
I don't like this option. It's too hard to get the splitting to be correct. I'll just make it check exec output.